### PR TITLE
Add type hints to test_descriptors.

### PR DESCRIPTION
**Pull request by clokep**
Builds on #15658 to finish typing our test files.

### CODE REVIEW

File: tests/util/caches/test_descriptors.py [hunk 1]
```diff
@@ -260,31 +271,31 @@ def fn(self, arg1):
...
                 return inner_fn()
 
         @defer.inlineCallbacks
-        def do_lookup():
+        def do_lookup() -> Generator["Deferred[Any]", object, int]:
             with LoggingContext("c1") as c1:
                 r = yield obj.fn(1)
                 self.assertEqual(current_context(), c1)
-            return r
+            return cast(int, r)
```
Comment (DMRobertson) [inline tests/util/caches/test_descriptors.py:296]: does `r: int` work here? I guess not if the send-type is `object`.

File: tests/util/caches/test_descriptors.py [hunk 2]
```diff
@@ -304,23 +315,23 @@ def check_result(r):
 
         return defer.gatherResults([d1, d2])
 
-    def test_cache_logcontexts_with_exception(self):
+    def test_cache_logcontexts_with_exception(self) -> Deferred:
```
Comment (DMRobertson) [inline tests/util/caches/test_descriptors.py:318]: Are you using `Deferred` without an inner type to mean a Deferred that is never finished with a value>

**Review by DMRobertson — APPROVED** [review]
I'm happy if mypy is!

File: tests/util/caches/test_descriptors.py [hunk 3]
```diff
@@ -260,31 +271,31 @@ def fn(self, arg1):
...
         """Check that logcontexts are set and restored correctly when
         using the cache."""
 
-        complete_lookup: Deferred = defer.Deferred()
+        complete_lookup: Deferred = Deferred()
 
         class Cls:
             @descriptors.cached()
-            def fn(self, arg1):
+            def fn(self, arg1: int) -> Deferred[int]:
```
Comment (DMRobertson) [inline tests/util/caches/test_descriptors.py:282]:
```suggestion
            def fn(self, arg1: int) -> "Deferred[int]":
```

File: tests/util/caches/test_descriptors.py (hunk 2 above)
Comment (DMRobertson) [reply tests/util/caches/test_descriptors.py:318]: > It essentially is Deferred[Any] I think.

Yep, I think any omitted types for a generic type are assumed to be Any.

Comment (clokep) [reply tests/util/caches/test_descriptors.py]: I was able to make this one into `Deferred[None]` but a couple other spots run into annoying things with Deferreds needing to change types when they go through callbacks and that not really being supported with the type system.

**Pull request merged**